Testscripts/Linux/perf_apache.sh

```diff
@@ -0,0 +1,255 @@
+#!/bin/bash
+
+#######################################################################
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache License.
+#
+#######################################################################
+
+#######################################################################
+#
+# perf_apache.sh
+# Description:
+#    Install apache and run apache performance tests.
+#    This script needs to be run on client VM.
+#
+# Supported Distros:
+#    Ubuntu/Suse/Centos/RedHat/Debian
+#######################################################################
+
+CONSTANTS_FILE="./constants.sh"
+UTIL_FILE="./utils.sh"
+
+. ${CONSTANTS_FILE} || {
+	errMsg="Error: missing ${CONSTANTS_FILE} file"
+	LogErr "${errMsg}"
+	SetTestStateAborted
+	exit 10
+}
+. ${UTIL_FILE} || {
+	errMsg="Error: missing ${UTIL_FILE} file"
+	LogErr "${errMsg}"
+	SetTestStateAborted
+	exit 10
+}
+
+if [ ! "${server}" ]; then
+	errMsg="Please add/provide value for server in constants.sh. server=<server ip>"
+	LogErr "${errMsg}"
+	SetTestStateAborted
+	exit 1
+fi
+if [ ! "${client}" ]; then
+	errMsg="Please add/provide value for client in constants.sh. client=<client ip>"
+	LogErr "${errMsg}"
+	SetTestStateAborted
+	exit 1
+fi
+
+if [ ! "${nicName}" ]; then
+	errMsg="Please add/provide value for nicName in constants.sh. nicName=eth0/bond0"
+	LogErr "${errMsg}"
+	SetTestStateAborted
+	exit 1
+fi
+
+localaddress=$(hostname -i)
+
+Run_SSHCommand()
+{
+	ips="$1"
+	cmd="$2"
+	IFS=',' read -r -a array <<< "$ips"
+	for ip in "${array[@]}"
+	do
+		LogMsg "Execute ${cmd} on ${ip}"
+		if [[ ${localaddress} = ${ip} ]]; then
+			bash -c "${cmd}"
+		else
+			ssh "${ip}" "${cmd}"
+		fi
+	done
+}
+
+# Install apache on client and server Machine
+LogMsg "Configuring client ${client}..."
+install_apache
+if [ $? -ne 0 ]; then
+	LogMsg "Error: apache installation failed in ${client}.."
+	SetTestStateAborted
+	exit 1
+fi
+
+LogMsg "Configuring server ${server}..."
+Run_SSHCommand "${server}" ". $UTIL_FILE && install_apache"
+if [ $? -ne 0 ]; then
+	LogMsg "Error: apache installation failed in ${server}.."
+	SetTestStateAborted
+	exit 1
+fi
+
+GetDistro
+case "$DISTRO_NAME" in
+	oracle|rhel|centos)
+		web_server="httpd"
+		document_path="/var/www/html/test.dat"
+		;;
+
+	ubuntu|debian)
+		web_server="apache2"
+		document_path="/var/www/html/test.dat"
+		;;
+
+	sles|sle_hpc)
+		web_server="apache2"
+		document_path="/srv/www/htdocs/test.dat"
+		;;
+	*)
+		LogErr "Unsupported distribution for apache test"
+		SetTestStateSkipped
+		exit 1
+esac
+
+mpstat_cmd="mpstat"
+dstat_cmd="dstat"
+sar_cmd="sar"
+vmstat_cmd="vmstat"
+
+testName="apache"
+log_folder="$(pwd)/${testName}-test-logs"
+max_concurrency_per_ab=4
+max_ab_instances=16
+testConcurrency=(1 2 4 8 16 32 64 128 256 512 1024)
```

Comment:
can we make this a replaceable parameter?
how long does it take to complete these tests?

Reply:
I see the email "LISA test case best practice" sent by Simon, "we should be able to run the test case itself (not in LISA) directly without considering other dependencies". So I put the parameter here instead of replaceableTestParameter.xml.

Reply:
It will take about 40 mins.

Reply:
setting parameter as replaceableTestParameter enables the flexibility to override settings from parameter.
You can set a default value if the parameter is not set, so that it can still run without LISAv2

Reply:
I have fixed this.